File: protocol/src/ptt-control.ts
export const PTT_CONTROL_TOPIC = "koeon.ptt.control";
export const PTT_CONTROL_FAST_START_TOPIC = "koeon.ptt.control.fast-start.v1";
export const PTT_CONTROL_VERSION = 1 as const;

export type PttControlType = "start" | "end";

export interface PttControlEvent {
  version: typeof PTT_CONTROL_VERSION;
  type: PttControlType;
  channelId: string;
  speakerUserId: string;
  sessionId: string;
  leaseId: string;
  sequence: number;
  sentAt: number;
}

export function encodePttControlEvent(event: PttControlEvent): Uint8Array<ArrayBuffer> {
  const encoded = new TextEncoder().encode(JSON.stringify(event));
  const copy = new Uint8Array(new ArrayBuffer(encoded.byteLength));
  copy.set(encoded);
  return copy;
}

export function decodePttControlEvent(data: Uint8Array): PttControlEvent | null {
  try {
    const value: unknown = JSON.parse(new TextDecoder().decode(data));
    if (!value || typeof value !== "object") return null;
    const event = value as Partial<PttControlEvent>;
    if (
      event.version !== PTT_CONTROL_VERSION
      || (event.type !== "start" && event.type !== "end")
      || !nonEmpty(event.channelId)
      || !nonEmpty(event.speakerUserId)
      || !nonEmpty(event.sessionId)
      || !nonEmpty(event.leaseId)
      || !Number.isSafeInteger(event.sequence)
      || (event.sequence ?? -1) < 0
      || !Number.isFinite(event.sentAt)
      || (event.sentAt ?? -1) < 0
    ) return null;
    return event as PttControlEvent;
  } catch {
    return null;
  }
}

function nonEmpty(value: unknown): value is string {
  return typeof value === "string" && value.length > 0 && value.length <= 256;
}

File: android/app/src/main/java/com/dennomuso/koeon/core/ptt/PttControl.kt
package com.dennomuso.koeon.core.ptt

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

const val PTT_CONTROL_TOPIC = "koeon.ptt.control"
const val PTT_CONTROL_FAST_START_TOPIC = "koeon.ptt.control.fast-start.v1"
const val PTT_CONTROL_VERSION = 1

@Serializable
data class PttControlEvent(
    val version: Int = PTT_CONTROL_VERSION,
    val type: String,
    val channelId: String,
    val speakerUserId: String,
    val sessionId: String,
    val leaseId: String,
    val sequence: Long,
    val sentAt: Long,
    val bufferedGenerationId: String? = null,
)

interface PttControlGateway {
    fun prepareRxReady(leaseId: String, expectedSessionIds: List<String>) = Unit
    fun prepareRxReady(leaseId: String, expectedSessionIds: List<String>, expectedDeviceIds: List<String>) =
        prepareRxReady(leaseId, expectedSessionIds)
    suspend fun publishStart(leaseId: String): Result<Unit>
    suspend fun publishBufferedStart(leaseId: String, generationId: String): Result<Unit> = publishStart(leaseId)
    suspend fun awaitRxReady(leaseId: String): RxReadyResult = RxReadyResult(
        reason = RxReadyReason.NO_EXPECTATIONS,
        expectedCount = 0,
        receivedCountAtMicOn = 0,
        ratioAtMicOn = 1.0,
        lateCount = 0,
        waitMs = 0,
    )
    suspend fun publishEnd(leaseId: String): Result<Unit>
    fun cancelRxReady() = Unit
}

object NoopPttControlGateway : PttControlGateway {
    override suspend fun publishStart(leaseId: String): Result<Unit> = Result.success(Unit)
    override suspend fun publishEnd(leaseId: String): Result<Unit> = Result.success(Unit)
}

internal val pttControlJson = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
}
